Add AoP docs to SN Hub. (#108)
streamnative/pulsar-hub: protocol-handlers/aop/0.1.0/aop.md

support_link: https://streamnative.io
support_img: "/images/connectors/streamnative.png"
owner_name: "StreamNative"
owner_img: "/images/streamnative.png"
dockerfile: 
id: "aop"
---

# Overview

AoP (AMQP on Pulsar) supports AMQP0-9-1 protocol and it is backed by Pulsar. AoP is implemented as a Pulsar [ProtocolHandler](https://github.com/apache/pulsar/blob/master/pulsar-broker/src/main/java/org/apache/pulsar/broker/protocol/ProtocolHandler.java) with the protocol name of "amqp". The AoP protocol handler is built as a `nar` file and it is loaded when the Pulsar broker starts.

# Installation

To install the AoP protocol handler, follow these steps:

1. Download the NAR package of the AoP protocol handler from [here](https://github.com/streamnative/aop/releases/download/v0.1.0/pulsar-protocol-handler-amqp-0.1.0-SNAPSHOT.nar).

2. Install the AoP protocol handler.

   1. Set the configuration of the AoP protocol handler.
       
       Add the following properties and set their values in the Pulsar configuration file, such as `conf/broker.conf` or `conf/standalone.conf`.

       Property | Recommended value | Default value
       |---|---|---
       `messagingProtocols` | amqp | null

>        `protocolHandlerDirectory`| Location of AoP NAR file | ./protocols

Reviewer: ```suggestion
       `protocolHandlerDirectory`| ./protocols | null
```

       
       **Example**

       ```
       messagingProtocols=amqp
       protocolHandlerDirectory=./protocols
       ```
       

   2. Set the AMQP server listeners.

       > #### Note
       >
       > The hostname in the AMQP server listeners should be the same one as the Pulsar broker's `advertisedAddress`.

       **Example**

       ```
       amqpListeners=amqp://127.0.0.1:5672
       advertisedAddress=127.0.0.1
       ```

3. Load the AoP protocol handler. After installing the AoP protocol handler on the Pulsar broker, you can restart the Pulsar broker to load the AoP protocol handler.

# Configuration 

The following table lists configurations available for the AoP protocol handler.

| Item | Description | Default value |
|---|---|---|
| amqpTenant | Set the AMQP on the Pulsar broker tenant. | public |
| amqpListeners | Set the AMQP service port. | amqp://127.0.0.1:5672 |

> | maxNoOfChannels | Set the maximum number of channels available concurrently on a connection. | 64 |

Reviewer: ```suggestion
| amqpMaxNoOfChannels | Set the maximum number of channels available concurrently on a connection. | 64 |
```

> | maxFrameSize | Set the maximum frame size on a connection. | 4 MB |

Reviewer: ```suggestion
| amqpMaxFrameSize | Set the maximum frame size on a connection. | 4194304 (4MB) |
```

> | heartBeat | Set the default heartbeat timeout on the Pulsar broker. | 60 s |

Reviewer: ```suggestion
| amqpHeartBeat | Set the default heartbeat timeout on the Pulsar broker. | 60 (s) |
```

> | amqpProxyPort | Set the AMQP proxy service port. | Port 5682 |

Reviewer: ```suggestion
| amqpProxyPort | Set the AMQP proxy service port. | 5682 |
```

2. Install the AoP protocol handler.

   1. Set the configuration of the AoP protocol handler.
       
       Add the following properties and set their values in the Pulsar configuration file, such as `conf/broker.conf` or `conf/standalone.conf`.

       Property | Recommended value | Default value
       |---|---|---
       `messagingProtocols` | amqp | null
       `protocolHandlerDirectory`| Location of AoP NAR file | ./protocols
       
       **Example**

       ```
       messagingProtocols=amqp
       protocolHandlerDirectory=./protocols
       ```
       

   2. Set the AMQP server listeners.

       > #### Note
       >
       > The hostname in the AMQP server listeners should be the same one as the Pulsar broker's `advertisedAddress`.

       **Example**

       ```
       amqpListeners=amqp://127.0.0.1:5672
       advertisedAddress=127.0.0.1
       ```

3. Load the AoP protocol handler. After installing the AoP protocol handler on the Pulsar broker, you can restart the Pulsar broker to load the AoP protocol handler.

    ```
    $PULSAR_HOME/bin/pulsar-daemon start standalone
    ```

4. Add the namespace for vhost.

    ```
    # for example, the vhost name is `vhost`

>     bin/pulsar-admin namespaces create public/vhost1 -b 1

Reviewer: ```suggestion
    bin/pulsar-admin namespaces create -b 1 public/vhost1
```
